### DISCUSSION TITLE
Adding select on LengthAwarePaginator confuses PHPStan

### DISCUSSION THREAD

**Discussion by minasm** [larastan/larastan#2067] [Q&A] [answered]
I know I can ignore this issue, but I am curious about how to make this work—after a few trials, I decided to see if I was missing something.

`paginate()` actually returns a [`LengthAwarePaginator`](https://laravel.com/api/11.x/Illuminate/Contracts/Pagination/LengthAwarePaginator.html) which extends [`AbstractPaginator`](https://laravel.com/api/11.x/Illuminate/Pagination/AbstractPaginator.html). I can see `AbstractPaginator` has [`withQueryString()`](https://laravel.com/api/11.x/Illuminate/Pagination/AbstractPaginator.html#method_withQueryString).

I have prepared a basic query below. The first query does not have `select()`, and it passes the checks.

```php
use App\Models\User;

$users  = User::orderBy('id','desc')->paginate(20)->withQueryString();
```

however, when I add a `select()` to the mix,

```php
use App\Models\User;

$users  = User::select('name')->orderBy('id','desc')->paginate(20)->withQueryString();
```
PHPStan complains about an undefined. 
```
Call to an undefined method Illuminate\Contracts\Pagination\LengthAwarePaginator::withQueryString(). 
```
How would you be able to solve this? I admit, it's not a big one, but it kept me thinking for a while now. 

I am using 
```
PHP 8.3
laravel/framework 11.27.2
larastan/larastan 2.9.8
phpstan/phpstan 1.12.1 (through Larastan)
phpstan/phpstan-mockery 1.1.2  
```

Thanks.

**Comment by calebdw** [comment]
I'm not able to reproduce, this passes just fine:

```php
assertType('Illuminate\Pagination\LengthAwarePaginator<App\User>', User::select('name')->orderBy('id', 'desc')->paginate(20)->withQueryString());
```

**Reply by calebdw** [reply]
It's picking up the contract (and the method is not defined on the contract). Do you actually have Larastan enabled? What does your `phpstan.neon(.dist)` look like?

**Reply by minasm** [reply]
Hello @calebdw,

Thanks for your reply!

Yes, I am using Larastan. Here is my `phpstan.neon` file;
```neon
includes:
    - vendor/larastan/larastan/extension.neon
    - vendor/phpstan/phpstan-mockery/extension.neon

parameters:

    paths:
        - app/

    # Level 9 is the highest level
    level: 6

#    ignoreErrors:
#        - '#PHPDoc tag @var#'
#
#    excludePaths:
#        - ./*/*/FileToBeExcluded.php
#
#    checkMissingIterableValueType: false
```

However, after your message, I got curious, so I created a new Laravel project (copied the same `composer.json` and `phpstan.neon`) and added these two different codes as different methods. Guess what? I did not get the error!

Now, I'm more curious about the issue, and I'll dig a bit deeper to see if I can find out.

**Reply by szepeviktor** [reply, marked as answer]
Try removing the vendor/ directory and clearing PHPStan's cache.

**Comment by minasm** [comment]
@szepeviktor, thank you so much! I removed the `vendor` directory and cleared the PHPStan cache, but I still have the same issue.
Then, I removed `composer.lock` and started over. Now, I do not get the issue.

I guess it's up to me now to analyse and see which package was the culprit.
